k2/python/k2/fsa_algo.py

```diff
@@ -145,3 +164,100 @@ def arc_sort(fsa: Fsa) -> Fsa:
     for name, value in fsa.named_non_tensor_attr():
         setattr(out_fsa, name, deepcopy(value))
     return out_fsa
+
+
+def compose(a_fsa: Fsa, b_fsa: Fsa) -> Fsa:
+    need_arc_map = True
+    ragged_arc, c_aux_labels, a_arc_map, b_arc_map = _k2.compose(
+        a_fsa=a_fsa.arcs,
+        b_fsa=b_fsa.arcs,
+        a_aux_labels=a_fsa.aux_labels.to(torch.int32),
+        b_aux_labels=b_fsa.aux_labels.to(torch.int32),
+        need_arc_map=need_arc_map)
+
+    # Some of entries in a_arc_map and b_arc_map may be zero.
+    # We add 1 here so that every entry is non-negative.
+    a_arc_map = a_arc_map.to(torch.int64) + 1
+    b_arc_map = b_arc_map.to(torch.int64) + 1
+
+    out_fsa = Fsa.from_ragged_arc(ragged_arc)
+    for name, a_value in a_fsa.named_tensor_attr():
+        if name == 'aux_labels':
+            # aux_label is handled specially
+            continue
+        if hasattr(b_fsa, name):
+            b_value = getattr(b_fsa, name)
+
+            # a_arc_map and b_arc_map have been offset by 1
+            # so we need a padding here
+            padding = a_value.new_zeros((1, *a_value.shape[1:]))
+            a_value = torch.cat((padding, a_value), dim=0)
+            b_value = torch.cat((padding, b_value), dim=0)
+
+            assert a_value.dtype == b_value.dtype == torch.float32
+            value = a_value.index_select(0, a_arc_map) \
+                    + b_value.index_select(0, b_arc_map)
+            setattr(out_fsa, name, value)
+
+    if hasattr(b_fsa, 'aux_labels'):
+        b_aux_labels = b_fsa.aux_labels
+        padding = b_aux_labels.new_zeros(1)
+        b_aux_labels = torch.cat((padding, b_aux_labels), dim=0)
+        out_fsa.aux_labels = b_aux_labels.index_select(0, b_arc_map)
+
+    for name, a_value in a_fsa.named_non_tensor_attr():
+        if name == 'osym':
+            continue
+        setattr(out_fsa, name, a_value)
+
+    for name, b_value in b_fsa.named_non_tensor_attr():
+        if not hasattr(out_fsa, name):
+            setattr(out_fsa, name, b_value)
+
+    return out_fsa
+
+
+def shortest_distance(fsa: Fsa) -> torch.Tensor:
```

Comment:
Just a note: when we compute the total-score in the log semiring, there won't be a way to do it as a pure function, we'll need a python Function object to implement the backprop.

We need to do a good job of indicating, in function documentation, what properties are required of the inputs.  For instance, here, does the Fsa have to have 2 axes (single FSA) and be on CPU?

Reply:
> when we compute the total-score in the log semiring, there won't be a way to do it as a pure function, we'll need a python Function object to implement the backprop.

GTN uses `log_semiring` in the crf example and I changed it to `tropical semiring` in k2.

I think the backprop for `log_semiring` has to be implemented in C++ and wrapped to
Python.

Reply:
Will update the documentation.

Reply:
The code needed for log-semiring backprop already exists, it involves GetForwardScores(), GetBackwradScores(), GetTotScores() to get the forward-pass output; and GetArcScores() [you'd have to exponentiate that to get the derivative].  Needs to be done as a Function in PyTorch.

Reply:
BTW, I think the Python-level Fsa should have a cache, indexed by string, which can store attributes like the batches and the forward and backward scores.  Some of them, like the forward and backward scores, should be stored with names that indicate which semiring we used, e.g. 'forward_log' vs. 'forward_tropical'.  This way we can re-use things where necessary.

Reply:
I have added two attributes: `arc_labels_sorted` and `arc_aux_labels_sorted` to avoid unnesseary `arc_sort` when
the fsa is already arc sorted.